### ISSUE TITLE
R8 rules for Californium

### ISSUE THREAD

**Issue by Thomas-Vos** [eclipse/californium#1583]
# Please provide basic issue information
* org.eclipse.californium:californium-core:2.6.2
* org.eclipse.californium:scandium:2.6.2
* R8 version 2.1.75 (on Android)

These are the R8 rules I am using for Californium 2.6.2 to get it working on Android with R8 enabled:
```
-keep class org.eclipse.californium.elements.util.StringUtil
-keep class org.slf4j.MDC { *; }
```

# -keep class org.eclipse.californium.elements.util.StringUtil
If this rule is not added you get the following exception. I deobfuscated the stack trace so that could be why it looks a bit strange. I also removed references to my own code.
```
java.lang.ExceptionInInitializerError
    at org.eclipse.californium.elements.util.StringUtil.normalizeLoggingTag(Unknown Source:0)
    at ***.org.eclipse.californium.scandium.config.DtlsConnectorConfig$Builder.build(SourceFile:28)
           setEndpoint
    at ***.<init>(SourceFile:14)
           <init>
    at ***(SourceFile:16)
    at ***.invokeSuspend(SourceFile:7)
    at ***.invoke(SourceFile:2)
           invoke
    at kotlin.time.DurationKt.kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(SourceFile:2)
    at kotlin.time.DurationKt.kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(SourceFile:9)
                              kotlinx.coroutines.BuildersKt.withContext
    at ***(SourceFile:1)
    at ***$suspendImpl(SourceFile:35)
    at ***(SourceFile:22)
       invokeSuspend
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(SourceFile:2)
    at kotlinx.coroutines.DispatchedTask.run(SourceFile:20)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(SourceFile:1)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(SourceFile:11)
                                                               runWorker
                                                               run
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Package.getImplementationVersion()' on a null object reference
    at org.eclipse.californium.elements.util.StringUtil.<clinit>(SourceFile:6)
    at org.eclipse.californium.elements.util.StringUtil.normalizeLoggingTag(Unknown Source:0)
    at ***.org.eclipse.californium.scandium.config.DtlsConnectorConfig$Builder.build(SourceFile:28)
           setEndpoint
    at ***.<init>(SourceFile:14)
           <init>
    at ***(SourceFile:16)
    at ***.invokeSuspend(SourceFile:7)
    at ***.invoke(SourceFile:2)
           invoke
    at kotlin.time.DurationKt.kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(SourceFile:2)
    at kotlin.time.DurationKt.kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(SourceFile:9)
                              kotlinx.coroutines.BuildersKt.withContext
    at ***(SourceFile:1)
    at ***$suspendImpl(SourceFile:35)
    at ***(SourceFile:22)
       invokeSuspend
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(SourceFile:2)
    at kotlinx.coroutines.DispatchedTask.run(SourceFile:20)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(SourceFile:1)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(SourceFile:11)
                                                               runWorker
                                                               run
```

It looks like something is going wrong here: 

https://github.com/eclipse/californium/blob/b12e60803324f68bdb0dd40175c8fbb29a454fc4/element-connector/src/main/java/org/eclipse/californium/elements/util/StringUtil.java#L85

# -keep class org.slf4j.MDC { *; }
I needed to add this rule after upgrading from Californium 2.2.3 to 2.6.2. Without the rule you get the following stack trace (deobfuscated):
```
MDC binding unsuccessful.
Reported exception:
java.lang.NullPointerException
    at org.slf4j.MDC.<clinit>(SourceFile:1)
    at org.eclipse.californium.scandium.DTLSConnector$6.doWork(SourceFile:1)
    at org.eclipse.californium.scandium.DTLSConnector$6.run(SourceFile:6)
```
It was hard for me to figure out what was going wrong, but I was able to "fix" the issue with the mentioned rule.

---

Most libraries for Android have R8 rules included which is really helpful. Maybe that could be done for Californium as well. Or if the issues above are fixed then that is not needed. I am not sure if any of the other Californium modules need rules for R8 as I do not use those.

**Comment by boaks** [comment]
> Most libraries for Android have R8 rules included which is really helpful. 

Sure, but I' m simply not to common to R8 :-).
There is a very simple [android demo app](https://github.com/eclipse/californium/tree/master/demo-apps/cf-android).  If you contribute your "gradle R8" build setup, I can check that in the future.

```java 
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Package.getImplementationVersion()' on a null object reference
    at org.eclipse.californium.elements.util.StringUtil.<clinit>(SourceFile:6)
```

So, your not interested in reading the Californium version :-).  Could you please check, if

```java 
String version = null;
Package pack = StringUtil.class.getPackage();
if (pack != null) {
	version = pack.getImplementationVersion();
}
CALIFORNIUM_VERSION = version;
```

works?

```java
MDC binding unsuccessful.
Reported exception:
java.lang.NullPointerException
    at org.slf4j.MDC.<clinit>(SourceFile:1)
    at org.eclipse.californium.scandium.DTLSConnector$6.doWork(SourceFile:1)
    at org.eclipse.californium.scandium.DTLSConnector$6.run(SourceFile:6)
```

OK, I guess I can check, if MDC is available and use it only, if so. 
Without the logging messages may lose their context information in the future. That was at least the plan when starting to use MDC.

**Comment by boaks** [comment]
I prepared a first PR for the "package". It will be included in 2.6.3 and in the 3.0.0-M1.

For the MDC I will check it the next days.

**Comment by Thomas-Vos** [comment]
The Android sample is really old, I am trying to update it (need to do that to use the latest R8). 

I cannot get the local server working. As soon as I tap Run Local Server it throws the following exception:

```
2021-03-29 21:21:25.715 29707-29707/org.eclipse.californium.examples.android E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.eclipse.californium.examples.android, PID: 29707
    java.lang.RuntimeException: Unable to create service org.eclipse.californium.examples.ServerService: java.lang.IllegalArgumentException: no credentials found for 'server' in 'classpath://certs/client.p12'!
        at android.app.ActivityThread.handleCreateService(ActivityThread.java:4601)
        at android.app.ActivityThread.access$1700(ActivityThread.java:301)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2149)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:246)
        at android.app.ActivityThread.main(ActivityThread.java:8456)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
     Caused by: java.lang.IllegalArgumentException: no credentials found for 'server' in 'classpath://certs/client.p12'!
        at org.eclipse.californium.elements.util.SslContextUtil.loadCredentials(SslContextUtil.java:459)
        at org.eclipse.californium.examples.ConfigureDtls.loadCredentials(ConfigureDtls.java:60)
        at org.eclipse.californium.examples.ServerService.onCreate(ServerService.java:44)
        at android.app.ActivityThread.handleCreateService(ActivityThread.java:4589)
        at android.app.ActivityThread.access$1700(ActivityThread.java:301) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2149) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:246) 
        at android.app.ActivityThread.main(ActivityThread.java:8456) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130) 

```

Is there anything I missed setting up the sample or so? This makes it a bit hard to test the code with R8 in the sample app. Unfortunately I am not familiar with the server parts as I only use Californium as a client in my app.

**Comment by boaks** [comment]
The server uses a certificate, which is in 'classpath://certs/client.p12'. That's a functionality in `SslContextUtil` to include x509 certificates as resource direct into the classpath. Alternatively it must be copied into android resources,. Or I adapt the code to use that optional (means, no x509, but PSK will work).

**Comment by boaks** [comment]
PSK_MODE is fine for me. It's more about the gradle stuff then the californium code. (The Californium code may be adapted by me later.)

**Comment by Thomas-Vos** [comment]
See the following: https://github.com/eclipse/californium/pull/1585

**Comment by boaks** [comment]
The `StringUtil` and `MDC` work-arounds are also merged to the 2.6.x and will be released as 2.6.3 in a couple of weeks. Or, if someone requests it, maybe earlier.
